crypto/ecdsa/ecdsa.go

```diff
@@ -127,225 +123,146 @@ func (sig ThresholdSignature) RangeWhile(f func(hotstuff.ID) bool) {
-	sig, ok := agg.(ThresholdSignature)
-	if !ok {
-		return false
-	}
-	results := make(chan bool)
-	for _, pSig := range sig {
-		go func(sig *Signature) {
-			results <- ec.mods.Crypto().Verify(sig, hash)
-		}(pSig)
+// Combine combines multiple signatures into a single signature.
+func (ec *ecdsaBase) Combine(signatures ...consensus.QuorumSignature) (consensus.QuorumSignature, error) {
+	if len(signatures) < 2 {
+		return nil, crypto.ErrCombineMultiple
 	}
-	valid := true
-	for range sig {
-		if <-results {
-			valid = false
-		}
-	}
-	return valid
-}
-
-// CreateThresholdSignature creates a threshold signature from the given partial signatures.
-func (ec *ecdsaCrypto) CreateThresholdSignature(partialSignatures []consensus.Signature, hash consensus.Hash) (_ consensus.ThresholdSignature, err error) {
-	thrSig := make(ThresholdSignature)
-	for _, s := range partialSignatures {
-		if thrSig.Participants().Contains(s.Signer()) {
-			err = multierr.Append(err, crypto.ErrPartialDuplicate)
-			continue
-		}
 
-		sig, ok := s.(*Signature)
-		if !ok {
-			err = multierr.Append(err, fmt.Errorf("%w: %T", crypto.ErrWrongType, s))
-			continue
-		}
+	ts := make(MultiSignature)
 
-		// use the registered verifier instead of ourself to verify.
-		// this makes it possible for the signatureCache to work.
-		if ec.mods.Crypto().Verify(s, hash) {
-			thrSig[sig.signer] = sig
+	for _, sig1 := range signatures {
+		if sig2, ok := sig1.(MultiSignature); ok {
+			for id, s := range sig2 {
+				if _, ok := ts[id]; ok {
+					return nil, crypto.ErrCombineOverlap
+				}
+				ts[id] = s
+			}
+		} else {
+			ec.mods.Logger().Panicf("cannot combine signature of incompatible type %T (expected %T)", sig1, sig2)
 		}
 	}
 
-	if len(thrSig) >= ec.mods.Configuration().QuorumSize() {
-		return thrSig, nil
-	}
-
-	return nil, multierr.Combine(crypto.ErrNotAQuorum, err)
+	return ts, nil
 }
 
-// CreateThresholdSignatureForMessageSet creates a ThresholdSignature of partial signatures where each partialSignature
-// has signed a different message hash.
-func (ec *ecdsaCrypto) CreateThresholdSignatureForMessageSet(partialSignatures []consensus.Signature, hashes map[hotstuff.ID]consensus.Hash) (_ consensus.ThresholdSignature, err error) {
-	ec.mods.Logger().Debug(hashes)
-	thrSig := make(ThresholdSignature)
-	for _, s := range partialSignatures {
-		if thrSig.Participants().Contains(s.Signer()) {
-			err = multierr.Append(err, crypto.ErrPartialDuplicate)
-			continue
-		}
+// Verify verifies the given quorum signature against the message.
+func (ec *ecdsaBase) Verify(signature consensus.QuorumSignature, message []byte) bool {
+	s, ok := signature.(MultiSignature)
+	if !ok {
+		ec.mods.Logger().Panicf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
+	}
 
-		hash, ok := hashes[s.Signer()]
-		if !ok {
-			continue
-		}
+	l := signature.Participants().Len()
```

Comment:
`s/l/n/g`

Reply:
The above comment meant: substitute `l` with `n`.

Reply:
Indeed, I am familiar with sed syntax ;)